File: postgresql/net9/Services/Test1Service.cs
using Microsoft.EntityFrameworkCore;
using net9.Context;
using net9.Models.Entities;
using net9.Models.Requests;
using net9.Models.Responses;

namespace net9.Services;

public class Test1Service: ITest1Service {
    private readonly PostgresContext _postgresContext;

    public Test1Service(PostgresContext postgresContext)
    {
        _postgresContext = postgresContext;
    }

    public async Task<Response<CreateResponse>> Create(CreateRequest createRequest)
    {
        using var transaction = await _postgresContext.Database.BeginTransactionAsync();
        try
        {
            var test1 = new Test1 {Test = createRequest.Test};
            await _postgresContext.Test1.AddAsync(test1);
            int rowsAffected = await _postgresContext.SaveChangesAsync();
            if (rowsAffected != 1) {
                return ResponseHelper.SetInternalServerErrorResponse<CreateResponse>();
            }
            await transaction.CommitAsync();
            return ResponseHelper.SetCreatedResponse<CreateResponse>(new CreateResponse{Id = test1.Id, Test = test1.Test});
        }
        catch(Exception e)
        {
            await transaction.RollbackAsync();
            Console.WriteLine($"error: {e}");
            return ResponseHelper.SetInternalServerErrorResponse<CreateResponse>();
        }
    }

    public async Task<Response<GetByIdResponse>> GetById(int id)
    {
        try
        {
            var test1 = await _postgresContext.Test1.AsNoTracking().FirstOrDefaultAsync(t => t.Id == id);
            if (test1 == null) {
                return ResponseHelper.SetNotFoundResponse<GetByIdResponse>($"cannot find test1 with id: {id}");
            }
            return ResponseHelper.SetOkResponse<GetByIdResponse>(new GetByIdResponse{Id = test1.Id, Test = test1.Test});
        }
        catch(Exception e)
        {
            Console.WriteLine($"error: {e}");
            return ResponseHelper.SetInternalServerErrorResponse<GetByIdResponse>();
        }
    }

    public async Task<Response<UpdateResponse>> Update(UpdateRequest updateRequest)
    {
        using var transaction = await _postgresContext.Database.BeginTransactionAsync();
        try
        {
            var test1 = await _postgresContext.Test1.FindAsync(updateRequest.Id);
            if (test1 == null) {
                return ResponseHelper.SetNotFoundResponse<UpdateResponse>($"cannot find test1 with id: {updateRequest.Id}");
            }
            test1.Test = updateRequest.Test;
            int rowsAffected = await _postgresContext.SaveChangesAsync();
            if (rowsAffected != 1) {
                return ResponseHelper.SetInternalServerErrorResponse<UpdateResponse>();
            }
            await transaction.CommitAsync();
            return ResponseHelper.SetOkResponse<UpdateResponse>(new UpdateResponse{Id = test1.Id, Test = test1.Test});
        }
        catch(Exception e)
        {
            await transaction.RollbackAsync();
            Console.WriteLine($"error: {e}");
            return ResponseHelper.SetInternalServerErrorResponse<UpdateResponse>();
        }
    }

    public async Task<Response<MessageResponse>> Delete(DeleteRequest deleteRequest)
    {
        using var transaction = await _postgresContext.Database.BeginTransactionAsync();
        try
        {
            var test1 = await _postgresContext.Test1.FindAsync(deleteRequest.Id);
            if (test1 == null) {
                return ResponseHelper.SetNotFoundResponse<MessageResponse>($"cannot find test1 with id: {deleteRequest.Id}");
            }
            _postgresContext.Remove(test1);
            var rowsAffected = await _postgresContext.SaveChangesAsync();
            if (rowsAffected != 1) {
                return ResponseHelper.SetInternalServerErrorResponse<MessageResponse>();
            }
            await transaction.CommitAsync();
            return ResponseHelper.SetNoContentResponse<MessageResponse>(null);

        }
        catch(Exception e)
        {
            await transaction.RollbackAsync();
            Console.WriteLine($"error: {e}");
            return ResponseHelper.SetInternalServerErrorResponse<MessageResponse>();
        }
    }
}

File: mongodb/rust/src/services/test1_service.rs
use std::sync::Arc;

use mongodb::bson::oid::ObjectId;

use crate::{models::{entities::test1::Test1, requests::{create_request::CreateRequest, delete_request::DeleteRequest, update_request::UpdateRequest}, responses::{create_response::CreateResponse, get_by_id_response::GetByIdResponse, response::Response, update_response::UpdateResponse}}, repositories::test1_repository::{Test1Repository, Test1RepositoryImpl}};

pub trait Test1Service {
    async fn create(&self, create_request: CreateRequest) -> Response<CreateResponse>;
    async fn get_by_id(&self, id: String) -> Response<GetByIdResponse>;
    async fn update_by_id(&self, update_request: UpdateRequest) -> Response<UpdateResponse>;
    async fn delete_by_id(&self, delete_request: DeleteRequest) -> Response<()>;
}

pub struct Test1ServiceImpl {
    test1_repository: Arc<Test1RepositoryImpl>
}

impl Test1ServiceImpl {
    pub fn new(test1_repository: Arc<Test1RepositoryImpl>) -> Test1ServiceImpl {
        Test1ServiceImpl {
            test1_repository
        }
    }
}

impl Test1Service for Test1ServiceImpl {
    async fn create(&self, create_request: CreateRequest) -> Response<CreateResponse> {
        let test1 = Test1 {id: ObjectId::new(), test: create_request.test};
        if let Err(err) = self.test1_repository.create(&test1).await {
            println!("error: {}", err);
            return Response::set_internal_server_error_response()
        }
        Response::set_created_response(CreateResponse::set_create_response(test1))
    }

    async fn get_by_id(&self, id: String) -> Response<GetByIdResponse> {
        let object_id = match ObjectId::parse_str(&id) {
            Ok(object_id) => object_id,
            Err(err) => {
                println!("error: {}", err);
                return Response::set_internal_server_error_response();
            }
        };
        let test1_option = match self.test1_repository.get_by_id(object_id).await {
            Ok(test1_option) => test1_option,
            Err(err) => {
                println!("error: {}", err);
                return Response::set_internal_server_error_response();
            }
        };

        let test1 = match test1_option {
            Some(test1) => test1,
            None => {
                println!("cannot find tes1 with id: {}", &id);
                return Response::set_not_found_response("cannot find test1 with id: ".to_owned() + &id);
            }
        };
        Response::set_ok_response(GetByIdResponse::set_get_by_id_response(test1))
    }

    async fn update_by_id(&self, update_request: UpdateRequest) -> Response<UpdateResponse> {
        let object_id = match ObjectId::parse_str(update_request.id) {
            Ok(object_id) => object_id,
            Err(err) => {
                println!("error: {}", err);
                return Response::set_internal_server_error_response();
            }
        };
        let test1 = Test1 { id: object_id, test: update_request.test };
        if let Err(err) = self.test1_repository.update_by_id(&test1).await {
            println!("error: {}", err);
            return Response::set_internal_server_error_response();
        }
        Response::set_ok_response(UpdateResponse::set_update_response(test1))
    }

    async fn delete_by_id(&self, delete_request: DeleteRequest) -> Response<()> {
        let object_id = match ObjectId::parse_str(delete_request.id) {
            Ok(object_id) => object_id,
            Err(err) => {
                println!("error: {}", err);
                return Response::set_internal_server_error_response();
            }
        };
        if let Err(err) = self.test1_repository.delete_by_id(object_id).await {
            println!("error: {}", err);
            return Response::set_internal_server_error_response();
        }
        Response::set_no_content_response()
    }
}